Write a user guide (#93)
samufi/larecs: docs/src/guide/entities_components_world.md

+++
type = "docs"
title = "Entities, components, and the world"
weight = 10
+++

> ## A little bit of introduction...

Reviewer: Would not start chapters with a level 2 heading, but with just a bit of introductory text. Particularly in this case, but also for the other chapters. It also looks a bit weird.

![grafik](https://github.com/user-attachments/assets/882caa19-c84a-4520-8053-e8464e39fc02)

Author: Okay, will do.

Larecs🌲 provides an efficient way to store, access and manipulate 

> data belonging to computational entities.

Reviewer: "computational" makes no sense here for me.

Author: What do we call entities if they have not been defined? How do we explain entity without using the word?

Reviewer: I think the entire chapter requires restructuring, which would resolve the issue. See [below](https://github.com/samufi/larecs/pull/93#discussion_r2023714128).

Author: Sure, but this was a genuine question: How do you explain entity without using the word?

Reviewer: Use the word entity, sure. Just "*computational* entity" makes no sense for me. Maybe for an HPC cluster or something, but not for ECS entities. What would that be, a "rechnerische Entität"?

Author: Please have a look at the updated version

Reviewer: The general structure and comprehensibility is much better now IMO. However, I still do not really like the introductory paragraph. Will try to come up with a suggestion tomorrow.

"Entities" often represent objects or agents that have certain
properties ("objects" in the object oriented framework),
here called "Components". Which attributes an entity has 
can change at runtime, making the approach extremely flexible.
Furthermore, entities do not belong to pre-defined 
"classes" but are rather fully characterized by their attributes.
A key functionality of entity component systems is to 
query all entities that have certain components and process them

> in a corresponding way. 

Reviewer: This entire first section is probably quite confusing for readers not familiar with ECS.

IMO the entire page would be more comprehensible if it is just structured like the title:

- Intro (no heading): 1-3 sentences about ECS (in general)
- Section "Entities":
  - What they represent: Game objects, "ODD entities", or potentially anything else that "lives" in the game or sim
  - What they are: IDs to access components (link next section)
- Section "Components":
  - What they represent: Entity data / state variables of entities
  - What they are: structs (or rather, any unique type?); (current) limitation trivial ...
- Section "The world"
  - central storage for all this stuff
  - entities are bound to a world
  -  ...

Author: Please have a look at the updated version.

For example, entities that can move may be characterized 
via the components `Position` and `Velocity`, as nothing
else is required for movement with constant velocity: 
the new position can be computed from the current position, 
the velocity, and a time increment `dt`:

```
Position = Position + Velocity * dt
```

Below, we provide more details on how entities 
and components are represented in Larecs🌲 and how
an ECS can be set up with Larecs🌲.

## Components

In Larecs🌲 and other ECS, components are modelled
via structs. That is, each component is represented
by a different struct. If we want to consider position 
and velocity as in the example above, we need to 
define a struct for each of the two:

```mojo {doctest="guide_entities_components_world" global=true}
@value
struct Position:
    var x: Float64
    var y: Float64

@value
struct Velocity:
    var dx: Float64
    var dy: Float64
```

> In this example, we consider a two-dimensional world.

Reviewer: Unnecessary detail that just extends the text, IMO. It is obvious.

Author: Removed.

the state of the entities and their surroundings.

Larecs🌲 gains a lot of its efficiency by using compile-time
programming. To that end, it needs to know which components
might turn up in the world at compile time, and the
`World` struct must be parameterized with the entities
upon creation. This also has the advantage that certain errors
can already be prevented at compile time, which makes the
program safer and faster.

To set up a `World`, simply import it from the `larecs` package
and create a `World` instance as follows:

```mojo {doctest="guide_entities_components_world" global=true}
from larecs import World

def main():
    world = World[Position, Velocity]()
```

## Entities

Entities are strictly bound to the world they live in.
An entity merely contains an ID that the world can use 
to look up the entity's components. As a consequence,
entities are small in memory and can be efficiently 
stored, and their data can only be accessed via the world.

> [!NOTE] 
> Though entities can safely be stored an passed around,
> their components (or pointers to them) should never be stored
> externally, as they can move in memory at any time.

> How entities are created and used will be discussed in the

Reviewer: Would always use present for cross refs: "is discussed".

You shouls also think about how to name "parts" of the docs. I use "chapter" for pages and "section" for sections (i.e. headings) of a chapter. Here, I would not know how to call sections, as it is already reserved.

Author: Ok, will do.

+++
type = "docs"
title = "Entities, components and the world"
weight = 10
+++

Larecs🌲 provides a flexible and efficient way to store, 
access and manipulate data records organized 
around individual entities. Suppose we wanted to 
model a world with different objects and agents that each have 
certain properties. The idea of entity component systems 
(ECS) is to characterize these objects and agents 
(called "entities" in an ECS) via the set of 
attributes ("components") they possess, and to 
model the interactions between them via "systems",
functions that operate on entities 
with certain components. Larecs🌲 provides the functionality 

> and tools to implement such entity component systems.

Reviewer: To be exact, it is not "an entity component system", i.e. it is not a system, but "entity component system architecture". If you want to call something the "system", than that's Larecs, not the thing you build with it.

So if you want to include such a statement, it should be something like "Larecs provides the functionality to build [something] following the ECS architecture".

But TBH, neither "Larecs🌲 provides a flexible and efficient way to store..." nor "Larecs🌲 provides the functionality..." are required here IMO. In my eyes, that's just fillers, as it is clear that we are talking about Larecs here.

Author: I await your suggestion for the intro section. I am very happy that you improve it!

## Entities

Entities are the central unit around which the data in
an ECS is organized. Entities could represent game objects,
or potentially anything else that "lives" in a game or simulation.
Each entity can possess an arbitrary
set of components, which can also be used to characterize
an entity. That is, in an ECS, cars would not directly 

> identified as "cars" but rather as everything that has 

Reviewer: ```suggestion
be identified as "cars" but rather as something that has 
```

Author: Done.

retrieving the corresponding components from the world.
As such, entities are strictly bound to the world they live in,
but also small in memory and easy to store.

> [!NOTE] 
> Though entities can safely be stored and passed around,
> their components (or pointers to them) should never be stored
> externally, as they can move in memory at any time.

How entities are created and used is discussed in the
[next chapters](../adding_and_removing_entities).

## Components

Components are the data associated with entities, characterizing
their state and properties. Components are represented
via structs: each component type is a different struct.
For example, if we want to model a world in which 
entities may have a position and a velocity, we need to 
define a `Position` and a `Velocity` struct.

```mojo {doctest="guide_entities_components_world" global=true}
@value
struct Position:
    var x: Float64
    var y: Float64

@value
struct Velocity:
    var dx: Float64
    var dy: Float64
```

These structs are not only used to store data. Their
types are also used as unique identifiers for the components
of an entity. That is, they have a similar role to 
a key in a dictionary.

> How components can be associated with entities is described 

Reviewer: IMO always writing "associated" makes things harder to understand. Although components are not stored in entities, I would prefer "can be added to entities".

> in the [later chapters](../adding_and_removing_entities).

Reviewer: Why not use concrete chapter titles?

Author: Because it is multiple chapters. Would you want me to list them all? I mainly want to tell the reader "hold on for now, it will become clear".

Reviewer: Ok, then maybe "following chapters"?

Author: Done.

> [!Warning]
> Currently, only "trivial" structs are supported as 
> components in Larecs🌲. That is, structs that can be
> copied and moved via simple memory operations. This does
> *not* include structs that manage heap-allocated memory
> such as `List` or `Dict`. Typically, it is not advisable to
> use such objects in the ECS context anyway; however, 
> Larecs🌲 might support such "complex" structs in a future version. 

## The world

> The central container type of Larecs🌲 is the

Reviewer: ```suggestion
The central container of Larecs🌲 is the
```

Author: Container "structure"? Otherwise one might think that larecs is contained...

Author: Or "the central data structure"

> {{< api World >}}. The `World` stores all data and information about

Reviewer: ```suggestion
{{< api World >}}. It stores all data and information about
```

Author: Done.

> the state of the entities and their surroundings.

Reviewer: Surroundings? This is probably very confising.

Author: I would like to write "the rest of the system", but "system" is a reserved word...

Reviewer: Maybe then "... stores entities, components and resources" (with a link on the latter)?
Or "... entities and their associated data"?

Author: Hmm, resources is difficult to understand before it has been introduced. I would like to find a different term for that. "entities and associated data" would sound to me like that each entity has associated data (which then would be components).

The `World` struct provides the main functionality
of the ECS, such as adding and removing entities,
looking up components, and iterating over entities.

> Larecs🌲 gains a lot of its efficiency by using compile-time

Reviewer: I do not agree here. Actually, I think it gains only the last small bit of performance from it, and the majority from the general architecture. But as you really seem to like this fact, go ahead... 😆

Author: "Larecs🌲 gains efficiency and usability by exploiting
Mojo's compile-time programming capabilities."?

+++
type = "docs"
title = "Entities, components and the world"
weight = 10
+++

> Larecs🌲 provides a flexible and efficient way to store, 
> access and manipulate data records organized 
> around individual entities. Suppose we wanted to 
> model a world with different objects and agents that each have 
> certain properties. The idea of entity component systems 
> (ECS) is to characterize these objects and agents 
> (called "entities" in an ECS) via the set of 
> attributes ("components") they possess, and to 
> model the interactions between them via "systems",
> functions that operate on entities 
> with certain components. Larecs🌲 provides the functionality 
> and tools to implement such entity component systems.

Reviewer: ```suggestion
Larecs🌲 is an entity component system (ECS) framework for Mojo.
The ECS concept is composed of three principal elements:

- [Entities](#entities) represent game objects or simulation entities, like individuals in a population model.
- [Components](#components) are the data associated to entities, i.e. their properties or state variables.
- [Systems](../systems_scheduler) contain the game or simulation logic that manipulates entities and their components, using so-called queries.

In an ECS, each entity is "composed of" an arbitrary set of components that can be added and removed at run-time.
This modular design enables the development of highly flexible and reusable games or simulations.
By decoupling the logic (systems) from the data (entities and components),
ECS avoids convoluted inheritance hierarchies and eliminates hard-coded behavior.
Instead, entities can be composed of components to exhibit diverse behaviors without sacrificing modularity.

An ECS engine manages this architecture within a central storage structure known as the "[World](#the-world)".
The engine handles common tasks such as maintaining entity lists, spawning or deleting entities,
and scheduling logic operations, simplifying the development process for users.

Larecs🌲 provides a high-performance ECS framework,
empowering developers to create games and simulation models with exceptional flexibility and efficiency.
```

For attribution, it would be nice if you just press "accept" (in case you agree) instead of incorporating the changes manually.
